tls/s2n_ktls_io.c

```diff
@@ -405,3 +406,35 @@ int s2n_sendfile(struct s2n_connection *conn, int in_fd, off_t offset, size_t co
     *blocked = S2N_NOT_BLOCKED;
     return S2N_SUCCESS;
 }
+
+int s2n_ktls_read_full_record(struct s2n_connection *conn, uint8_t *record_type)
+{
+    POSIX_ENSURE_REF(conn);
+    POSIX_ENSURE_REF(record_type);
+
+    /* This method copies data into conn->in, so is intended for control messages
+     * rather than application data. However in some cases-- such as when attempting
+     * to read the close_notify alert during s2n_shutdown-- it may encounter application
+     * data. Set a reasonable conn->in size to avoid an excessive number of calls
+     * to recvmsg when reading a larger record.
+     */
+    POSIX_GUARD(s2n_stuffer_resize_if_empty(&conn->in, S2N_DEFAULT_FRAGMENT_LENGTH));
```

Comment:
I'm not sure about this.

This basically means that we'll read control messages in chunks of S2N_DEFAULT_FRAGMENT_LENGTH, as if we'd received them in records with fragment length == S2N_DEFAULT_FRAGMENT_LENGTH. We do this because without the record header (which ktls consumes) we don't actually know how much data is expected / available.

A couple alternatives:
* I could initially set the buffer / fragment size to S2N_ALERT_LENGTH and resize larger if application data or handshake messages are encountered. That would mean an extra small recv for application data and handshake messages, but would optimize for alerts?
* I could make an initial recvmsg call with the MSG_PEEK | MSG_TRUNC flags set, which should let me know how much data is available (according to the documentation-- I haven't tested this with ktls). But this means that even in the happy case, where no application data is encountered, we'd make two system calls. It'd also likely make this implementation harder to consolidate with quic, because the recv callback has no mechanism to implement peek. MSG_TRUNC is also only available in later versions of Linux, but the same is true of ktls so that's probably not an issue.

Reply:
Oh now that is an interesting idea! In fact, we might be able to extend it to alerts and handshake messages too. conn->in usually needs to be allocated on the heap to handle the situation where the application provided output buffer isn't large enough to hold all the application data from a record, so we need to hold onto the rest until the next call. But if we're only interested in alerts and handshake messages, partial alerts are buffered in conn->alert_in and partial handshake messages are buffered in conn->post_handshake.in. So with ktls, conn->in should always be read completely before we call read_full_record again.

Let me play around with this and see if we can allocated conn->in on the stack for ktls. If it gets too complicated, I might make the change in a separate PR like I did for ktls send.

Reply:
Okay, so:
* Using stack memory for all messages: not possible without major changes, and not worth the complexity. If we want to only replace read_full_record, then we need to end up with data in conn->in where the rest of our code expects it. We could set conn->in to use stack memory, but that'd need to happen outside of read_full_record to be accessible outside of read_full_record.
* Using stack memory only to skip application data: not worth the complexity. To do this, we'd need to call recvmsg in a loop instead of just once. This significantly complicates this method, particularly because we don't know whether the next call to recvmsg actually will return more application data. If it doesn't, then we need to copy the control message back into conn->in. It also very arguably changes this from "read_full_record" to "read_multiple_full_records". Everyone calling this already calls it in a loop to process multiple records, so let's let that existing looping do the work.

Reply:
Actually, reverted for now. I think it'll make more sense if I optimize the memory for receiving app data and control messages at the same time. It might be controversial :)